eclair-core/src/test/scala/fr/acinq/eclair/payment/receive/InvoicePurgerSpec.scala

```diff
@@ -0,0 +1,91 @@
+      timestamp = 1 unixsec))
+    val expiredPayments = expiredInvoices.map( invoice => IncomingPayment(invoice, randomBytes32(), PaymentType.Standard, invoice.createdAt.toTimestampMilli, IncomingPaymentStatus.Expired))
+    expiredInvoices.lazyZip(expiredPayments).foreach((invoice, payment) => db.addIncomingPayment(invoice, payment.paymentPreimage))
+
+    // create pending invoices
+    val pendingInvoices = Seq.fill(count)(Bolt11Invoice(Block.TestnetGenesisBlock.hash, Some(100 msat), randomBytes32(), alicePriv, Left("pending invoice"), CltvExpiryDelta(18),
+      timestamp = TimestampSecond.now() - 600))
+    val pendingPayments = pendingInvoices.map( invoice => IncomingPayment(invoice, randomBytes32(), PaymentType.Standard, invoice.createdAt.toTimestampMilli, IncomingPaymentStatus.Pending))
+    pendingInvoices.lazyZip(pendingPayments).foreach((invoice, payment) => db.addIncomingPayment(invoice, payment.paymentPreimage))
+
+    // create paid invoices
+    val receivedAt = TimestampMilli.now() + 1.milli
+    val paidInvoices = Seq.fill(count)(Bolt11Invoice(Block.TestnetGenesisBlock.hash, Some(100 msat), randomBytes32(), alicePriv, Left("paid invoice"), CltvExpiryDelta(18),
+      timestamp = TimestampSecond.now()))
+    val paidPayments = paidInvoices.map( invoice => IncomingPayment(invoice, randomBytes32(), PaymentType.Standard, invoice.createdAt.toTimestampMilli, IncomingPaymentStatus.Received(100 msat, receivedAt)))
+    paidInvoices.lazyZip(paidPayments).foreach((invoice, payment) => { db.addIncomingPayment(invoice, payment.paymentPreimage)
+      // receive payment
+      db.receiveIncomingPayment(invoice.paymentHash, 100 msat, receivedAt)
+    })
+
+    val now = TimestampMilli.now()
+    assert(db.listIncomingPayments(0 unixms, now) === expiredPayments ++ pendingPayments ++ paidPayments)
+    assert(db.listIncomingPayments(expiredInvoices.last.createdAt.toTimestampMilli, now) === pendingPayments ++ paidPayments)
+    assert(db.listPendingIncomingPayments(0 unixms, now) === pendingPayments)
+    assert(db.listReceivedIncomingPayments(0 unixms, now) === paidPayments)
+    assert(db.listExpiredIncomingPayments(0 unixms, now) === expiredPayments)
+
+    val interval = 1 seconds
+
+    val _ = system.spawn(InvoicePurger(db, interval), name = "purge-expired-invoices")
+
+    val probe = TestProbe()
+    system.eventStream.subscribe(probe.ref, classOf[PurgeEvent])
+
+    // check that purge completed
+    probe.expectMsg(3 seconds, PurgeCompleted)
+    probe.expectNoMessage()
+    assert(db.listExpiredIncomingPayments(0 unixms, now).isEmpty)
+    assert(db.listIncomingPayments(0 unixms, now) === pendingPayments ++ paidPayments)
+
+    // add more expired invoices
+    expiredInvoices.lazyZip(expiredPayments).foreach((invoice, payment) => db.addIncomingPayment(invoice, payment.paymentPreimage))
```

Comment:
I now understand why you were concerned with race conditions, there is one here.
Since your invoice purger is running in the background, if it "ticks" before you've finished adding all these new payments your `probe.expectNoMessage()` won't work.

I don't think you need to add multiple invoices here for the test to be relevant. If you add a single new invoice (I would avoid reusing the previous `expiredInvoices`, it's a bit confusing) then this race condition disappears.

Reply:
fixed in commit [a38262f](https://github.com/ACINQ/eclair/pull/2174/commits/a38262f3f71b64fbc59beae0729d3a046cff222d) to create new expired invoices.

I forgot to make it  just a single 2nd expired invoice, but shouldn't hurt to do 10 of them either.

Reply:
Actually it does, it introduces a race condition because your purger actor is also running in parallel and ticking every second. The following scenario can happen:

- you start adding these 10 new invoices to the DB
- you've added 3 out of 10 but the `TickPurge` event happens inside the purger actor
- it will purge those 3 invoices and emit a `PurgeCompleted` event
- meanwhile you've added more invoices to the DB
- `TickPurge` will tick again and purge the other invoices, and emit another `PurgeCompleted` event

So your `probe` will receive two distinct purge events, which will make the `purge.expectNoMessage()` fail.
Since this is a race condition it will rarely happen, but will nonetheless sometimes make the CI build fail, which is a pain...